Enhance Documentation: Alternatives to `select!` Macro

Ddystopia opened the pull request:
This PR adds documentation for ecosystem alternatives to the `select!` macro, such as `tokio_stream::StreamExt::merge` and `futures-concurrency` utilities. Providing these alternatives helps users make informed decisions by showcasing additional tools and patterns, as a lot of new users are not aware of any alternatives to `select!`.

Ddystopia commented:
Are there some issues with pr?

Darksonn reviewed (commented):
Sorry, I forgot to click submit on my review.

My most general comment is that I'm not sure the selection of alternatives is complete. Surely we cannot discuss alternatives to the `tokio::select!` macro without mentioning the futures crate? It also has a macro called `select!` that is different from our macro in various ways.

More generally, Tokio usually links to futures over futures-lite because the futures crate is an official rust-lang project. As for futures-concurrency, I am okay with linking to that since the direct purpose of that crate is to investigate alternatives to select, so it is obvious to mention it.

Ddystopia commented:
Thank you for the review, I've changed examples, moved the whole section after `select!` examples and focused motivation on cancellation safety issues

Ddystopia commented:
Oh, there is an infinite loop in example 😅

Ddystopia commented:
@Darksonn I added `futures-concurrency` as dev-dependency in order to remove `ignore` from code snippet - is it acceptable?

Darksonn commented on tokio/src/macros/select.rs:
````diff
@@ -398,6 +398,151 @@ macro_rules! doc {
         /// ```
+        /// # Alternatives from the Ecosystem
+        ///
+        /// The select! macro is a powerful tool for managing multiple asynchronous
````
```suggestion
        /// The `select!` macro is a powerful tool for managing multiple asynchronous
```

Darksonn commented on tokio/src/macros/select.rs:
```diff
@@ -398,6 +398,151 @@ macro_rules! doc {
+        /// For cases where `loop { select! { ... } }` is used to poll multiple tasks,
+        /// stream merging offers a concise alternative, inherently handle cancellation-safe
+        /// processing, removing the risk of data loss. Libraries like
+        /// [`tokio_stream`](https://docs.rs/tokio-stream/latest/tokio_stream/),
+        /// [`futures::stream`](https://docs.rs/futures/latest/futures/stream/) and
+        /// [`futures_concurrency`](https://docs.rs/futures-concurrency/latest/futures_concurrency/)
```
Please double-check whether these links work without specifying the link. These are all available as dev-dependencies, so they might resolve without a https link.

If that doesn't work, then please move the address to a separate line:
```
Libraries such as [`tokio-stream`], [`futures::stream`], and [`futures-concurrency`]

[`tokio-stream`]: https://docs.rs/tokio-stream
```

Ddystopia replied:
Unfortunately, modules cannot be linked without https link. I will move addresses to a separate line

Darksonn commented on tokio/src/macros/select.rs:
```diff
@@ -398,6 +398,151 @@ macro_rules! doc {
+        ///         let data = socket.read_packet().await;
+        ///         Some((Message::Data(data), socket))
+        ///     });
+        ///     let c = iter([Message::Stop]);
```
I think this would be easier to read if it looks like this:
```suggestion
        ///     let c = stream::iter([Message::Stop]);
```

Ddystopia replied:
So import should be `use tokio_stream::{StreamExt, self as stream};`? Or write as `tokio_stream::iter`?

Darksonn commented on tokio/src/macros/select.rs:
```diff
@@ -398,6 +398,151 @@ macro_rules! doc {
+        /// [`futures-concurrency`](https://docs.rs/futures-concurrency/latest/futures_concurrency/)
+        /// provide streamlined syntax for racing futures:
+        ///
+        /// - [`futures_concurrency::future::Race`](https://docs.rs/futures-concurrency/latest/futures_concurrency/future/trait.Race.html)
+        /// - [`futures::select!`](https://docs.rs/futures/latest/futures/macro.select.html)
+        /// - [`futures::stream::select_all`](https://docs.rs/futures/latest/futures/stream/select_all/index.html) (for streams)
+        /// - [`futures_lite::future::or`](https://docs.rs/futures-lite/latest/futures_lite/future/fn.or.html)
+        /// - [`futures_lite::future::race`](https://docs.rs/futures-lite/latest/futures_lite/future/fn.race.html)
```
I suspect most of these links do not actually need an url.

Ddystopia replied:
I tried and all of them require a url. Only links like [`std::iter::once`] are resolved as is.

Darksonn reviewed (approved):
There's an unrelated CI breakage right now. I'll merge this when that is resolved.
